### PR TITLE
Support running macOS prebuilt application

### PR DESCRIPTION
**Pull request by jonahwilliams**
After building the app once with XCode, flutter run can accept the path to the app bundle. This automatically starts the app and attaches for hot reload & debugging.

Example:
`flutter run --use-application-binary ~/Library/Developer/Xcode/DerivedData/ExampleEmbedder-abcdefg/Build/Products/Debug/Example\ Embedder.app`

~Will wait on https://github.com/flutter/flutter/pull/26422 to land first to handle conflict in application_package.dart~ Landed

### CODE REVIEW

**Comment by jonahwilliams** [comment]
cc @stuartmorgan

File: packages/flutter_tools/lib/src/macos/macos_device.dart [hunk 1]
```diff
@@ -59,22 +62,38 @@ class MacOSDevice extends Device {
...
     bool usesTerminalUi = true,
     bool ipv6 = false,
-  }) {
-    throw UnimplementedError();
+  }) async {
+    if (!prebuiltApplication) {
+      return LaunchResult.failed();
+    }
+    // The location of these files is based on the examples in the flutter
+    // desktop embedding repo and may not be final.
```
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/macos/macos_device.dart]: The `Contents/MacOS` part is intrinsic to macOS apps, not a function of FDE.

It's also standard for the executable for `Foo.app` to be called `Foo`, but end users can rename the `.app` part at any time. The robust way to construct the path would be to get the `CFBundleExecutable` value from `Contents/Info.plist` (that path is also defined by macOS application bundle structure) and use that as the final path component.

File: packages/flutter_tools/lib/src/macos/application_package.dart [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+// found in the LICENSE file.
+
+import 'package:meta/meta.dart';
+
+import '../application_package.dart';
+import '../base/file_system.dart';
+import '../globals.dart';
+import '../ios/plist_utils.dart' as plist;
+
+ /// Tests whether a [FileSystemEntity] is an iOS bundle directory
```
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/macos/application_package.dart]: s/iOS/macOS/

File: packages/flutter_tools/lib/src/macos/application_package.dart [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        return null;
+      }
+      bundleDir = fs.directory(applicationBinary);
+    } else {
+      printError('Folder "${applicationBinary.path}" is not an app bundle.');
+      return null;
+    }
+    final String plistPath = fs.path.join(bundleDir.path, 'Contents', 'Info.plist');
+    if (!fs.file(plistPath).existsSync()) {
+      printError('Invalid prebuilt MacOS app. Does not contain Info.plist.');
```
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/macos/application_package.dart]: Nit: s/MacOS/macOS/ everywhere in strings.

**Review by stuartmorgan-g — COMMENTED** [review]
Nice :)

File: packages/flutter_tools/lib/src/macos/macos_device.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/macos/macos_device.dart]: Done

File: packages/flutter_tools/lib/src/macos/macos_device.dart [hunk 4]
```diff
@@ -20,19 +27,13 @@ class MacOSDevice extends Device {
   DeviceLogReader getLogReader({ApplicationPackage app}) => NoOpDeviceLogReader('macos');
 
   @override
-  Future<bool> installApp(ApplicationPackage app) {
-    throw UnimplementedError();
-  }
+  Future<bool> installApp(ApplicationPackage app) async => true;
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/macos/macos_device.dart]: Some comments about why these are hard-coded to true could be useful. Is there actually an implementation step that could be needed in the future? If so, then adding a TODO here as well could be a good pointer.

File: packages/flutter_tools/lib/src/macos/macos_device.dart [hunk 5]
```diff
@@ -59,26 +60,36 @@ class MacOSDevice extends Device {
...
+    } finally {
+      await observatoryDiscovery.cancel();
+    }
   }
 
   @override
-  Future<bool> stopApp(ApplicationPackage app) {
-    throw UnimplementedError();
-  }
+  Future<bool> stopApp(ApplicationPackage app) async => true;
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/macos/macos_device.dart]: Kill the process?

File: packages/flutter_tools/lib/src/macos/application_package.dart [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+import '../ios/plist_utils.dart' as plist;
+
+ /// Tests whether a [FileSystemEntity] is an macOS bundle directory
+bool _isBundleDirectory(FileSystemEntity entity) =>
+    entity is Directory && entity.path.endsWith('.app');
+
+abstract class MacOSApp extends ApplicationPackage {
+  MacOSApp({@required String projectBundleId}) : super(id: projectBundleId);
+
+   /// Creates a new [MacOSApp] from an existing app bundle.
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/macos/application_package.dart]: A little more detail here could be good. How is it going to do that?

File: packages/flutter_tools/lib/src/macos/macos_device.dart (hunk 4 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/macos/macos_device.dart]: Still working on this

File: packages/flutter_tools/lib/src/macos/macos_device.dart (hunk 5 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/macos/macos_device.dart]: Implemented

File: packages/flutter_tools/lib/src/macos/application_package.dart (hunk 6 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/macos/application_package.dart]: Done

File: packages/flutter_tools/lib/src/macos/macos_device.dart (hunk 4 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/macos/macos_device.dart]: Done

**Review by zanderso — APPROVED** [review]
lgtm

**Pull request merged**